### PR TITLE
chore(agents-api): slice docs to top k in case docs len > limit

### PR DESCRIPTION
**Pull request by Vedantsahai18**
### **User description**
---
# EntelligenceAI PR Summary 
 The updates in `gather_messages.py` and `search_docs.py` introduce a backup plan to retrieve the top `k` documents when embeddings are missing.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced `apply_mmr_to_docs` utility for Maximal Marginal Relevance (MMR).

- Refactored MMR logic across multiple modules for reusability.

- Added type checking using `beartype` for enhanced validation.

- Created comprehensive tests for MMR functionality and edge cases.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>get_doc_search.py</strong><dd><code>Add type checking and simplify `get_language`.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/common/utils/get_doc_search.py

<li>Added <code>beartype</code> type checking to functions.<br> <li> Simplified and validated <code>get_language</code> function.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1215/files#diff-6cd50924db8f0ca2e89a951db3be7cf2cae078d5c6d7d100d3abea2f0cdb79d3">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>mmr.py</strong><dd><code>Add reusable MMR utility with type checking.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/common/utils/mmr.py

<li>Introduced <code>apply_mmr_to_docs</code> for reusable MMR logic.<br> <li> Added type checking with <code>beartype</code>.<br> <li> Defined constants and helper types for clarity.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1215/files#diff-618f417b0650fb17e2e1c18fbfbc7beeb2f6dd6d0ea75fb5d72da97033a8f426">+47/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>gather_messages.py</strong><dd><code>Refactor MMR logic using utility function.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/queries/chat/gather_messages.py

<li>Replaced inline MMR logic with <code>apply_mmr_to_docs</code>.<br> <li> Improved readability and reusability of MMR application.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1215/files#diff-1b7f13a02e66bbfa4c5963a5d6339d096f0cfcf5cdaeb414fac26c1e6189f4a0">+9/-22</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>search_docs.py</strong><dd><code>Refactor document search MMR logic.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/routers/docs/search_docs.py

<li>Refactored MMR logic to use <code>apply_mmr_to_docs</code>.<br> <li> Simplified MMR application in document search endpoints.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1215/files#diff-7c78741623ef43e8990dbf7bc74bf97e64950e60d60eae1249afb33eb38083af">+15/-37</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_get_doc_search.py</strong><dd><code>Update tests for `get_language` with type validation.</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/test_get_doc_search.py

<li>Updated tests for <code>get_language</code> to handle <code>beartype</code> exceptions.<br> <li> Removed redundant test cases for invalid inputs.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1215/files#diff-8d711571115dd4b7f5da1b5a7f2ca7cbc39ed31c220a57465c88dc6b2ca3b151">+4/-15</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_mmr.py</strong><dd><code>Add tests for MMR utility and edge cases.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/test_mmr.py

<li>Added comprehensive tests for <code>apply_mmr_to_docs</code>.<br> <li> Covered edge cases like empty docs and varying MMR strengths.<br> <li> Validated behavior with and without embeddings.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1215/files#diff-67979b93c53f760e2f22198bd39a213a3eede52f8c0618dc5eb60cc31a4f90e9">+118/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

----

> [!IMPORTANT]
> Introduces `apply_mmr_to_docs` for MMR logic, refactors existing code to use it, and adds comprehensive tests.
> 
>   - **Behavior**:
>     - Introduced `apply_mmr_to_docs` in `mmr.py` for reusable MMR logic.
>     - Refactored MMR logic in `gather_messages.py` and `search_docs.py` to use `apply_mmr_to_docs`.
>     - Added type checking with `beartype` in `get_doc_search.py` and `mmr.py`.
>   - **Tests**:
>     - Added tests for `apply_mmr_to_docs` in `test_mmr.py`, covering edge cases like empty docs and varying MMR strengths.
>     - Updated tests in `test_get_doc_search.py` for `get_language` and search function parameters.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=julep-ai%2Fjulep&utm_source=github&utm_medium=referral)<sup> for bec402bdd9ee69b3c6b7a4434abf2701c09261fc. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: agents-api/agents_api/queries/chat/gather_messages.py [hunk 1]
```diff
@@ -179,5 +179,8 @@ async def gather_messages(
             doc_references = [
                 doc for i, doc in enumerate(docs_with_embeddings) if i in set(indices)
             ]
+        else:
+            # If there are no docs with embeddings, return the top k docs
```
Comment (whiterabbit1983) [inline agents-api/agents_api/queries/chat/gather_messages.py:183]: There's either comment is misleading or the logic is incorrect, because `MIN_DOCS_WITH_EMBEDDINGS` equals 2, so we can have 1 doc with embeddings, what should we do in this case?

File: agents-api/agents_api/queries/chat/gather_messages.py [hunk 2]
```diff
@@ -179,5 +179,8 @@ async def gather_messages(
...
+            doc_references = doc_references[: recall_options.limit]
```
Comment (whiterabbit1983) [inline agents-api/agents_api/queries/chat/gather_messages.py:184]: The same code chunk is repeated three times across the code base, I'd sugges to refactor it into a function and write a test for it

Comment (creatorrr) [reply agents-api/agents_api/queries/chat/gather_messages.py:184]: yeah the whole MMR thing is duplicated, should be just one function

Comment (Vedantsahai18) [reply agents-api/agents_api/queries/chat/gather_messages.py:184]: Yup, that's what my next thought step was.

File: agents-api/agents_api/queries/chat/gather_messages.py (hunk 1 above)
Comment (whiterabbit1983) [reply agents-api/agents_api/queries/chat/gather_messages.py:183]: I see, just update/remove this comment after you finish

Comment (Vedantsahai18) [reply agents-api/agents_api/queries/chat/gather_messages.py]: refactored into a function `common/utils/mmr.py`

File: agents-api/agents_api/queries/chat/gather_messages.py (hunk 2 above)
Comment (Vedantsahai18) [reply agents-api/agents_api/queries/chat/gather_messages.py]: refactored into a function common/utils/mmr.py. Also added typechecks using `beartype` and relevant tests.
